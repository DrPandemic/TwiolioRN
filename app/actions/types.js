// @flow

import { PhoneNumber, Message } from '../types';

export const SET_FETCHED_ACCOUNT_NUMBERS = 'SET_FETCHED_ACCOUNT_NUMBERS';
export const FETCH_ACCOUNT_NUMBERS = 'FETCH_ACCOUNT_NUMBERS';
export const FETCH_ACCOUNT_NUMBER_ERROR = 'FETCH_ACCOUNT_NUMBER_ERROR';

export const SELECT_NUMBER = 'SELECT_NUMBER';
export const SELECT_CONVERSATION = 'SELECT_CONVERSATION';

export const FETCH_MESSAGES = 'FETCH_MESSAGES';
export const SET_FETCHED_MESSAGES = 'SET_FETCHED_MESSAGES';
export const FETCH_MESSAGE_ERROR = 'FETCH_MESSAGE_ERROR';

export type FetchAccountNumbersT = {|
  type: 'FETCH_ACCOUNT_NUMBERS',
|}
export type SuccessFetchAccountNumbersT = {|
  type: 'SET_FETCHED_ACCOUNT_NUMBERS',
  fetchedAccountNumbers: Array<PhoneNumber>,
|}
export type FailFetchAccountNumbersT = {|
    type: 'FETCH_ACCOUNT_NUMBER_ERROR',
    error: any,
|}

export type SelectNumberT = {|
  type: 'SELECT_NUMBER',
  selectedNumber: ?string,
|}

export type FetchMessagesT = {|
  type: 'FETCH_MESSAGES',
|}
export type SuccessFetchMessagesT = {|
    type: 'SET_FETCHED_MESSAGES',
    fetchedMessages: Array<Message>,
|}
export type FailFetchMessagesT = {|
    type: 'FETCH_MESSAGE_ERROR',
    error: any,
|}

export type ActionT =
  | FetchAccountNumbersT
  | SuccessFetchAccountNumbersT
  | FailFetchAccountNumbersT
  | SelectNumberT
  | SelectConversationT
  | FetchMessagesT
  | SuccessFetchMessagesT
  | FailFetchMessagesT;
