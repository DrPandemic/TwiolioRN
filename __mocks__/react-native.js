const rn = require('react-native');
jest.mock('Linking', () => {
  return {
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    openURL: jest.fn(),
    canOpenURL: jest.fn(),
    getInitialURL: jest.fn(),
  };
});
module.exports = rn;
