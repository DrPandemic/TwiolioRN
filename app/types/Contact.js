// @flow

import {
  PhoneNumberFormat as PNF,
  PhoneNumberUtil
} from 'google-libphonenumber';

import { Message } from './';
import { flatMap } from '../lib/array';

const phoneUtil = PhoneNumberUtil.getInstance();

export type Email = {|
  label: string,
  email: string
|};

export type ContactPhoneNumber = {|
  label: string,
  number: string
|};

export type PostalAddress = {|
  postCode: string,
  city: string,
  neighborhood: string,
  street: string,
  formattedAddress: string,
  label: string
|};

export default class Contact {
  recordID: string;
  company: string;
  emailAddresses: Array<Email>;
  familyName: string;
  givenName: string;
  jobTitle: string;
  middleName: string;
  phoneNumbers: Array<ContactPhoneNumber>;
  hasThumbnail: boolean;
  thumbnailPath: string;
  postalAddresses: Array<PostalAddress>;

  constructor(response: any) {
    this.recordID = response.recordID;
    this.company = response.company;
    this.emailAddresses = response.emailAddresses;
    this.familyName = response.familyName;
    this.givenName = response.givenName;
    this.jobTitle = response.jobTitle;
    this.middleName = response.middleName;
    this.phoneNumbers = response.phoneNumbers.map(
      p => ({
        ...p,
        number: phoneUtil.format(phoneUtil.parse(p.number, 'US'), PNF.E164),
      })
    );
    this.hasThumbnail = response.hasThumbnail;
    this.thumbnailPath = response.thumbnailPath;
    this.postalAddresses = response.postalAddresses;
  }
}

export type PhoneNumberWithContact = {
  contact: Contact,
  phoneNumber: string
};

export function findContactsForNumber(
  contacts: Array<Contact>,
  number: string,
): Array<Contact> {
  const formattedNumber = phoneUtil.parse(number);
  return contacts.filter(
    c => c.phoneNumbers.filter(
      n => phoneUtil.isNumberMatch(phoneUtil.parse(n.number), formattedNumber) >= 2
    ).length > 0
  );
}

export function getNameForMessage(
  contacts: Array<Contact>,
  message: Message,
): string {
  const number = message.conversationUsers.other;
  const foundContacts = findContactsForNumber(contacts, number);

  return foundContacts.length > 0 ? foundContacts[0].givenName : number;
}

export function expandPhoneNumbers(contacts: Array<Contact>): Array<PhoneNumberWithContact> {
  return flatMap(contacts,
    contact => contact.phoneNumbers.map(
      n => ({ contact, phoneNumber: n.number })
    ));
}
