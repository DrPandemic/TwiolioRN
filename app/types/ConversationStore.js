// @flow

import { Message } from './';
import type { ConversationUsers } from './Message';
import type { NestedArray } from '../types';

export type ConversationStoreT = {
  [conversationId: string]: Array<Message>
};

// Should never produce an empty conversation.
export function addMessage(store: ConversationStoreT, message: Message):
ConversationStoreT {
  const id = message.conversationId;

  return {
    ...store,
    [id]: [message, ...(store[id] || []).filter(m => m.sid !== message.sid)]
  };
}

export function addMessages(
  store: ConversationStoreT,
  messages: Array<Message>
): ConversationStoreT {
  return messages.reduce((s, m) => addMessage(s, m), store);
}

export function getMessages(
  store: ConversationStoreT,
  conversationUsers: ConversationUsers
): Array<Message> {
  return store[Message.getConversationId(conversationUsers)];
}

export function getConversations(
  store: ConversationStoreT
): NestedArray<Message> {
  return Object.values(store);
}

// For this function to work, we need to have non-empty conversations.
export function filterByUs(
  store: ConversationStoreT,
  us: string
): ConversationStoreT {
  return Object.entries(store)
               .filter(
                 ([k, [v]]: [string, Array<Message>]) =>
                   v.conversationUsers.us === us)
               .reduce(
                 (acc, [k: string, v: Array<Message>]) => ({ ...acc, [k]: v }),
                 {}
               );
}
