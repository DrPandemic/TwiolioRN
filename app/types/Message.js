// @flow

export type ConversationUsers = {|
  us: string,
  other: string
|};

export default class Message {
  sid: string;
  body: string;
  to: string;
  from: string;
  dateCreated: string;
  dateUpdated: string;
  dateSent: string;
  status: string;
  direction: string;

  constructor(response: any) {
    this.sid = response.sid;
    this.body = response.body;
    this.to = response.to;
    this.from = response.from;
    this.dateCreated = response.date_created;
    this.dateUpdated = response.date_updated;
    this.dateSent = response.date_sent;
    this.status = response.status;
    this.direction = response.direction;
  }

  /*
     According to twilio doc:
     The direction of this SMS message. inbound for incoming messages,
      outbound-api for messages initiated via the REST API,
      outbound-call for messages initiated during a call
      or outbound-reply for messages initiated in response to an incoming SMS
   */
  get conversationUsers(): ConversationUsers {
    return {
      us: this.direction === 'inbound' ? this.to : this.from,
      other: this.direction === 'inbound' ? this.from : this.to
    };
  }

  static getConversationId(conversationUsers: ConversationUsers): string {
    return [conversationUsers.us, conversationUsers.other].join('');
  }

  get conversationId(): string {
    return Message.getConversationId(this.conversationUsers);
  }
}

