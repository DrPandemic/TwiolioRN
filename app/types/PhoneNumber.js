// @flow

export default class PhoneNumber {
  sid: string;
  friendlyName: string;
  number: string;

  constructor(response: any) {
    this.sid = response.sid;
    this.friendlyName = response.friendly_name;
    this.number = response.phone_number;
  }
}
