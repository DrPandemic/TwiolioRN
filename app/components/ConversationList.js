// @flow

import React, { Component } from 'react';
import { View, StyleSheet } from 'react-native';
import { List, ListItem } from 'react-native-elements';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import { ActionCreators } from '../actions';
import { Message } from '../types';
import { getConversations } from '../types/ConversationStore';
import type { StateT } from '../reducers';
import type { T as AccountT } from '../reducers/account';
import type { T as MessagesT } from '../reducers/messages';

const styles = StyleSheet.create({
  container: {
    flex: 1,
    marginTop: 30,
  },
});

type PropsT = {
  account: AccountT,
  messages: MessagesT,
}

export class PConversationList extends Component {
  props: PropsT;

  static renderRow([message]: Array<Message>) {
    return (
      <ListItem
        key={message.conversationId}
        title={message.conversationUsers.other}
      />
    );
  }

  render() {
    return (
      <View style={styles.container}>
        <List>
          {getConversations(this.props.messages.messages)
            .map(c => PConversationList.renderRow(c))}
        </List>
      </View>
    );
  }
}

function mapStateToProps(state: StateT) {
  return {
    account: state.account,
    messages: state.messages,
  };
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators(ActionCreators, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(PConversationList);
