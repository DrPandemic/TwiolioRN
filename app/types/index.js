// @flow
import Message from './Message';
import PhoneNumber from './PhoneNumber';
import ConversationStore from './ConversationStore';

export {
  Message,
  PhoneNumber,
  ConversationStore,
};
