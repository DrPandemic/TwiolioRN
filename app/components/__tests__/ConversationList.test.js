// @flow

import React from 'react';
import { View } from 'react-native';
import renderer from 'react-test-renderer';

import { PConversationList } from '../ConversationList';
import messageFixture from '../../test_helpers/fixtures/received_message.json';
import { Message } from '../../types';
import { addMessages, getConversations } from '../../types/ConversationStore';

test('renders a row', () => {
  const store = addMessages({}, [
    new Message(messageFixture.simple),
    new Message(messageFixture.simpleInverse),
    new Message(messageFixture.simpleOtherFrom),
    new Message(messageFixture.simpleOutbound),
  ]);
  const row = getConversations(store)[0];

  expect(
    renderer.create(PConversationList.renderRow(row))
  ).toMatchSnapshot();
});

test('renders rows', () => {
  const store = addMessages({}, [
    new Message(messageFixture.simple),
    new Message(messageFixture.simpleInverse),
    new Message(messageFixture.simpleOtherFrom),
    new Message(messageFixture.simpleOutbound),
  ]);

  expect(
    renderer.create(<View>{PConversationList.renderRows(store)}</View>)
  ).toMatchSnapshot();
});
