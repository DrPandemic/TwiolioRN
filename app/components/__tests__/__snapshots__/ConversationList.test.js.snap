// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a row 1`] = `
<View
  onLongPress={undefined}
  onPress={undefined}
  style={
    Array [
      Object {
        "backgroundColor": "transparent",
        "borderBottomColor": "#ededed",
        "borderBottomWidth": 1,
        "paddingBottom": 10,
        "paddingRight": 10,
        "paddingTop": 10,
      },
      undefined,
    ]
  }
  underlayColor="white"
>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "marginLeft": 10,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <View
        style={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {},
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              Array [
                Object {
                  "color": "#43484d",
                  "fontSize": 17.5,
                },
                Object {
                  "marginLeft": 10,
                },
                undefined,
                undefined,
              ],
            ]
          }
        >
          +1231231235
        </Text>
      </View>
      <View
        style={undefined}
      >
        <View />
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "flex-end",
          "flex": 0.15,
          "justifyContent": "center",
        }
      }
    >
      <View
        onPress={undefined}
        style={
          Array [
            false,
            false,
            false,
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "justifyContent": "center",
            },
            undefined,
          ]
        }
        underlayColor="white"
      >
        <Text
          accessible={true}
          allowFontScaling={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#bdc6cf",
                "fontSize": 28,
              },
              Array [
                Object {
                  "backgroundColor": "transparent",
                },
                undefined,
              ],
              Object {
                "fontFamily": "Material Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
            ]
          }
        >
          î—Œ
        </Text>
      </View>
    </View>
  </View>
</View>
`;
